src/vsphere_cpi/lib/cloud/vsphere/cloud.rb

```diff
@@ -662,6 +662,50 @@ def info
       {'stemcell_formats' =>  ['vsphere-ovf', 'vsphere-ova']}
     end
 
+    #creates T1 router and virtual switch attached to it
+    def create_subnet(subnet_definition)
+      raise 'NSXT must be enabled in CPI to use create_subnet' if !@config.nsxt_enabled?
+      cloud_properties = subnet_definition['cloud_properties']
+      raise 'cloud_properties must be provided' if cloud_properties.nil?
+      raise 'edge_cluster_id cloud property can not be empty' if cloud_properties['edge_cluster_id'].nil?
+      raise 't0_router_id cloud property can not be empty' if cloud_properties['t0_router_id'].nil?
+      raise 'transport_zone_id cloud property can not be empty' if cloud_properties['transport_zone_id'].nil?
+      subnet = create_subnet_obj(subnet_definition['range'], subnet_definition['gateway'])
+
+      t1_router_id = nil
+      switch_id = nil
+      begin
+        t1_router = @nsxt_provider.create_t1_router(cloud_properties['edge_cluster_id'], cloud_properties['t1_name'])
+        t1_router_id = t1_router.id
+        @nsxt_provider.enable_route_advertisement(t1_router_id)
+        @nsxt_provider.attach_t1_to_t0(cloud_properties['t0_router_id'], t1_router_id)
+        switch = @nsxt_provider.create_logical_switch(cloud_properties['transport_zone_id'], cloud_properties['switch_name'])
+        switch_id = switch.id
+        @nsxt_provider.attach_switch_to_t1(switch_id, t1_router_id, subnet)
+      rescue Exception => e
```

Comment:
Rescuing `Exception` is almost never what you want. We should be specific about the possible error types here.

Reply:
We need to clean up what we created. Also exception thrown by nsxt client are empty. So I think it's nice to have a meaningful message.